### PR TITLE
Feature/sensor observation, imu reward functions

### PR DESCRIPTION
**Pull request by vybhav-ibr**
adds the rate limited sensor observation function, and reward functions which use the imu measurement, makes some modifications to the observation manager to get one specific observation reading from the observation buffer

### CODE REVIEW

File: genesis_forge/mdp/observations.py [hunk 1]
```diff
@@ -205,3 +206,36 @@ def contact_force(env: GenesisEnv, contact_manager: ContactManager) -> torch.Ten
...
+
+
+"""
+Rate limited sensors
+"""
+
+
+class SensorObservation(MdpFnClass):
+    """
+    Helper class for arte limited Sensor observations
```
Comment (jgillick) [inline genesis_forge/mdp/observations.py]: "arte limited" -> "rate limited"

File: genesis_forge/mdp/observations.py [hunk 2]
```diff
@@ -205,3 +206,36 @@ def contact_force(env: GenesisEnv, contact_manager: ContactManager) -> torch.Ten
...
     """
     return torch.norm(contact_manager.contacts[:, :, :], dim=-1)
+
+
+"""
+Rate limited sensors
+"""
+
+
+class SensorObservation(MdpFnClass):
```
Comment (jgillick) [inline genesis_forge/mdp/observations.py]: This is not really limited to sensor readings -- it executes a read function that could return anything. Instead, it should likely be called something like "RateLimitedObservation".

File: genesis_forge/mdp/rewards.py [hunk 3]
```diff
@@ -406,6 +407,78 @@ def stand_still_joint_deviation_l1(
     return joint_deviation * (torch.norm(command[:, :2], dim=1) < command_threshold)
 
 
+"""
+Imu
+"""
+
+
+def imu_lin_acc_jitter(
```
Comment (jgillick) [inline genesis_forge/mdp/rewards.py:415]: Rewards should not be tied to the observation manager, as the readings might be scaled or noise might be added. Instead of relying on IMU data, what's wrong with the approach used by `lin_vel_z_l2`?

File: genesis_forge/mdp/rewards.py [hunk 4]
```diff
@@ -406,6 +407,78 @@ def stand_still_joint_deviation_l1(
...
+    _env: GenesisEnv,
+    observation_manager: ObservationManager,
+    obs_item_key: str,
+    ignore_gravity: bool = False,
+) -> torch.Tensor:
+    """
+    Penalize changes in linear acceleration (jerk) using the IMU's internal queue.
```
Comment (jgillick) [inline genesis_forge/mdp/rewards.py:422]: How is this using the "IMU's internal queue"? it appears to be querying the observation manager, not the IMU.

File: genesis_forge/mdp/rewards.py [hunk 5]
```diff
@@ -406,6 +407,78 @@ def stand_still_joint_deviation_l1(
...
+    ]
+    if ignore_gravity:
+        lin_acc_buffer[:, :, 2] += 9.81
+
+    if lin_acc_buffer.shape[0] > 1:
+        diffs = lin_acc_buffer[:, 1:] - lin_acc_buffer[:, :-1, :]
+    else:
+        print(
+            "the observation history buffer only has only the sensor data for one timsteps, the reward will be zero"
+        )
```
Comment (jgillick) [inline genesis_forge/mdp/rewards.py]: This seems like it would be printed every training session, and there's nothing for the developer to do with this warning. Let's avoid these kind of messages to the terminal.

File: genesis_forge/mdp/reset.py [hunk 6]
```diff
@@ -291,3 +291,67 @@ def __call__(
             links_idx_local=self._links_idx_local,
             envs_idx=envs_idx,
         )
+
+
+class randomize_link_com_shift(ResetMdpFnClass):
```
Comment (jgillick) [inline genesis_forge/mdp/reset.py]: This appears to be an exact duplicate of `randomize_link_mass_shift`. What's the purpose of adding this in this PR? Since it's unrelated to the subject of the PR, please remove it.

**Review by jgillick — CHANGES_REQUESTED** [review]

File: genesis_forge/mdp/rewards.py (hunk 3 above)
Comment (vybhav-ibr) [reply genesis_forge/mdp/rewards.py:415]: lin_vel may not always be observable in the real world(as it would require robot state estimation using using SLAM, Odometry, Ground truth sensor, GPS, etc), so this could be an alternative ?
I think this will help with sim2real, if the IMU data is part of the observations to the RL policy,
Should i shift this to a prototype PR ?, or remove it entirely ?

File: genesis_forge/mdp/rewards.py (hunk 4 above)
Comment (vybhav-ibr) [reply genesis_forge/mdp/rewards.py:422]: when i was using individual sensor manger, I was buffering for the IMU reading, now i am using the history buffer from the observation. I am not satisfied with this approach, as the size of history buffer would have to be increased to increase the IMU observation horizon, but this increase the size of the policy NN, If you would like to include the IMU rewards, i will create a class and buffer the IMU reading there, open to suggestions of any other way to do this

File: genesis_forge/mdp/reset.py (hunk 6 above)
Comment (vybhav-ibr) [reply genesis_forge/mdp/reset.py]: `randomize_link_mass_shift` modifies the mass of the link, `randomize_link_com_shift` modifies the center of mass of the link with respect to the link frame. I forgot to modify the `__call__` operator. i will make the changes and shift to a different PR

File: genesis_forge/mdp/rewards.py (hunk 3 above)
Comment (jgillick) [reply genesis_forge/mdp/rewards.py:415]: Please remove this entirely. Rewards should not be dependent on observation manager processed values, as they might be scaled or noise added. Rewards based on sensor readings should read the values from the sensor directly.

File: examples/sensors/trained.mov
Comment (jgillick) [inline examples/sensors/trained.mov:1]: Is this duplicated from the original example, or was it created during your training run?

File: examples/sensors/README.md
Comment (jgillick) [inline examples/sensors/README.md:1]: This file was directly copied without updates. The README should reflect what the example is demonstrating. Never blindly copy things and submit them for review without updating them appropriately.

File: genesis_forge/mdp/rewards.py [hunk 7]
```diff
@@ -406,6 +407,75 @@ def stand_still_joint_deviation_l1(
...
+
+    if lin_acc_buffer.shape[0] > 1:
+        diffs = lin_acc_buffer[:, 1:] - lin_acc_buffer[:, :-1, :]
+    else:
+        return torch.zeros(_env.num_envs, 1)
+    mags = torch.norm(diffs, dim=-1)
+    return mags.sum(dim=1)
+
+
+def imu_ang_vel_jitter(
```
Comment (jgillick) [inline genesis_forge/mdp/rewards.py:447]: Same comments as `imu_lin_acc_jitter`

**Review by jgillick — CHANGES_REQUESTED** [review]
